### DISCUSSION TITLE
nginx reverse proxy로 docker blue green 무중단 배포 하기

### DISCUSSION THREAD

**Discussion by minhyeok0328** [minhyeok0328/next-loop#88] [General]
# 무중단 배포
- 서비스가 새로 배포될 때 사용자가 불편을 겪지 않게 하기 위해 무중단 배포를 해야 합니다.

# 설명
1. 우선 인스턴스에 있는 nginx가 80 트래픽을 받습니다.
2. 그 후에 8000, 8001 포트에서 실행중인 2개의 FastAPI 서버로(Docker Container) 트래픽을 전달해 줄 겁니다
3. 두 인스턴스를 적절히 스위칭 해주면서 nginx에서 8000포트에서 실행중인 기존 컨테이너 에서 새로 배포된 8001 컨테이너로 트래픽을 전환시켜 줌으로써 사용자는 중단없이 서비스를 사용할 수 있게 됩니다.

> 참고: [Nginx Reverse Proxy Document](https://docs.nginx.com/nginx/admin-guide/web-server/reverse-proxy/)
# 환경 세팅
> Ubuntu를 기준으로 설명합니다.
## Nginx 설치
```
sudo apt update
sudo apt install -y nginx
```

## 설정파일 수정

설정 파일을 열고
`sudo vim /etc/nginx/sites-available/default`

아래 내용 대로 추가합니다.
```
# nginx 설정파일 /etc/nginx/sites-available/default

# upstream에 blue green app 정의해줍니다. 각각 8000, 8001 포트로 해두고
upstream blue_app { 
        server 127.0.0.1:8000;
}

upstream green_app {
        server 127.0.0.1:8001;
}

server {
        listen 80 default_server;
        listen [::]:80 default_server;
        root /var/www/html;
        index index.html index.htm index.nginx-debian.html;
        server_name _;
        location / {
                proxy_pass http://blue_app; ## 우리는 여기 문자열을 blue_app 또는 green_app으로 스위칭 해줘야 합니다
        }
}
```
추가로
`try_files $uri $uri/ =404;`
이 부분을 지워주셔야 하위 uri도 proxy 됩니다

예를 들면
`http://{ip}/docs`

이런 것들이요

## sh 파일 생성
switch 하는 과정을 자동화 해주기 위해 sh 파일을 작성합니다.
해당 파일은 아래 설명과 같은 역할을 합니다.

1. 기존에 동작중인 blue 컨테이너가 있습니다.
2. 새로운 파일을 업데이트 해야하니 docker image를 빌드합니다.
3. 그 후 새 컨테이너(green)를 생성합니다.
4. sed 명령어로 nginx 설정파일에 있는 설정을 변경해 줍니다. (proxy_pass http://blue_app; > proxy_pass http://green_app;)
5. nginx 설정파일을 reload 해줍니다
6. 마지막으로 기존 컨테이너(blue)를 종료합니다

```
#!/bin/bash

NGINX_CONFIG="/etc/nginx/sites-available/default"
DOCKER_COMMAND_HEAD="docker compose -f docker-compose.production.yml"
CURRENT_ENV=$(grep "proxy_pass" $NGINX_CONFIG | grep -o "blue_app\|green_app")

# blue green switch할 변수 추가
if [ "$CURRENT_ENV" = "blue_app" ]; then
    NEW_ENV="green_app"
else
    NEW_ENV="blue_app"
fi

# 이미지 새로 빌드
sudo $DOCKER_COMMAND_HEAD build
sudo $DOCKER_COMMAND_HEAD restart $NEW_ENV

# 혹시 모르니까 30초 정도 지연
sleep 30

# nginx proxy_pass 값 blue green switch
sudo sed -i "/proxy_pass/s/$CURRENT_ENV/$NEW_ENV/" $NGINX_CONFIG

# switch 하고 nginx 테스트 문제 없으면 switch 아니면 원래대로
if sudo nginx -t; then
    sudo nginx -s reload
    echo "Switched from $CURRENT_ENV to $NEW_ENV and reloaded Nginx."
else
    echo "Nginx configuration test failed. Rolling back changes."
    sudo sed -i "/proxy_pass/s/$NEW_ENV/$CURRENT_ENV/" $NGINX_CONFIG
fi

# 여기도 충분히 트래픽 전환될 시간 주기 위해 잠시 지연
sleep 10

# 기존 컨테이너는 stop
sudo $DOCKER_COMMAND_HEAD stop $CURRENT_ENV
```
특정 컨테이너 이름을 지정해서 컨테이너를 종료하거나 시작할 수 있습니다
예를 들면
```
services:
  blue_app:
    build:
      context: .
      dockerfile: Dockerfile
    container_name: blue_app
    ...
```
이런 container가 있을 때
sudo docker compose stop **blue_app** < 이렇게요

이렇게 하면 간단하게 무중단 배포가 완성 됩니다.

**Comment by imella0707** [comment]
이런 방식은 처음 알게 되었어요. blure, green도  민혁님이 그냥 이름 지은건줄 알았는데 관용적으로 쓰는 용어였군요. 감사합니다 !
